src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php

```diff
@@ -72,154 +84,86 @@ protected function configure()
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $this->writeIntroduction($output);
+        $output = new SymfonyStyle($input, $output);
+
+        $input->isInteractive() ? $output->title('Symfony Password Encoder Utility') : $output->newLine();
 
         $password = $input->getArgument('password');
-        $salt = $input->getArgument('salt');
         $userClass = $input->getArgument('user-class');
+        $emptySalt = $input->getOption('empty-salt');
 
-        $helper = $this->getHelper('question');
+        $encoder = $this->getContainer()->get('security.encoder_factory')->getEncoder($userClass);
+        $bcryptWithoutEmptySalt = !$emptySalt && $encoder instanceof BCryptPasswordEncoder;
+
+        if ($bcryptWithoutEmptySalt) {
+            $emptySalt = true;
+        }
 
         if (!$password) {
+            if (!$input->isInteractive()) {
+                $output->error('The password must not be empty.');
+
+                return 1;
+            }
             $passwordQuestion = $this->createPasswordQuestion($input, $output);
-            $password = $helper->ask($input, $output, $passwordQuestion);
+            $password = $output->askQuestion($passwordQuestion);
         }
 
-        if (!$salt) {
-            $saltQuestion = $this->createSaltQuestion($input, $output);
-            $salt = $helper->ask($input, $output, $saltQuestion);
-        }
+        $salt = null;
+
+        if ($input->isInteractive() && !$emptySalt) {
+            $emptySalt = true;
 
-        $output->writeln("\n <comment>Encoders are configured by user type in the security.yml file.</comment>");
+            $output->note('The command will take care about generating a salt for you. Be aware that some encoders advise to let them generate their own salt. If you\'re using one of those encoders, please answer \'no\' to the question below. '.PHP_EOL.'Provide the \'empty-salt\' option in order to let the encoder handle the generation itself.');
```

Comment:
**Very** minor point: "The command will generate a salt for you" would be more simple and more correct (otherwise "take care of") :smile: